pages/badges/badge-single.jsx

```diff
@@ -0,0 +1,449 @@
+            <textarea rows={10} onChange={this.updateEvidenceText} value={this.state.evidenceText} placeholder="Describe what you've done to earn this badge..."/>
+          </fieldset>
+
+          <fieldset>
+            <label className="control-label">Attach an (optional) link as part of your evidence:</label>
+            <input type="text" placeholder="Give a link to a page to act as evidence" value={this.state.evidenceLink} onChange={this.updateEvidenceLink} />
+          </fieldset>
+
+          <fieldset>
+            <input type="file" className="hidden" ref="optionalFile" onChange={this.handleFiles}/>
+            <label className="control-label">Attach an (optional) file as part of your evidence:</label>
+            <button className="btn attach" onClick={this.selectFiles}>pick file...</button>
+          </fieldset>
+
+          <button className={"btn"} disabled={!showButton} onClick={this.claimBadge}>Apply</button>
+        </div>
+      </div>
+    );
+  },
+
+  updateEvidenceText: function(evt) {
+    this.setState({
+      evidenceText: evt.target.value
+    });
+  },
+
+  updateEvidenceLink: function(evt) {
+    this.setState({
+      evidenceLink: evt.target.value
+    });
+  },
+
+  selectFiles: function() {
+    this.refs.optionalFile.click();
+  },
+
+  handleFiles: function(evt) {
+    var component = this;
+    var files = evt.target.files;
+    var attachments = [];
+    Array.from(files).forEach(function(file) {
+      var reader = new FileReader();
+      reader.onload = (function(f) {
+        return function(evt) {
+          var name = escape(f.name);
+          var data = evt.target.result;
+          if (data) {
+            // FIXME: TODO: There is a 20MB limit on file uploads whch
+            //              I doubt we'll run into, but _might_ be an issue
+            //              eventually, so I'm leaving this comment.
+            data = data.substring(data.indexOf('base64,')+'base64,'.length);
+            attachments.push({ name: name, file: data });
+          }
+
+          if(attachments.length === files.length) {
+            component.setState({
+              evidenceFiles: attachments
+            });
+          }
+        };
+      })(file);
+      reader.readAsDataURL(file);
+    });
+  },
+
+  claimBadge: function() {
+    var evidences = [];
+
+    if (this.state.evidenceText) {
+      evidences.push({
+        file: btoa(this.state.evidenceText),
+        name: "Claim description"
+      });
+    }
+
+    if (this.state.evidenceLink) {
+      evidences.push({
+        file: this.state.evidenceLink,
+        name: "Proof of claim link"
+      });
+    }
+
+    if (this.state.evidenceFiles.length > 0) {
+      evidences = evidences.concat(this.state.evidenceFiles);
+    }
+
+    if (evidences.length === 0) {
+      return console.error("a badge claim without evidence was attempted");
+    }
+
+    this.setState({
+      applying: true,
+      showApplyModal: true
```

Comment:
I think we should probably update the `showApplyModal` state in `this.handleClaimRequest` instead.  Things like network problem could fail the form submission and showing `Thanks for applying for this badge!` modal regardless the submission state could confuse users.

STR: 
1. go to https://teach-with-credly-badges.herokuapp.com/badge/1901/problem-solving
2. fill out the application form (without submitting it just yet)
3. do NOT interact with anything on the site for maybe 30 mins(?) for the app to go to sleep
4. hit "Apply" button to submit the badge application form
5. you are now being presented a "Thanks for applying for this badge!" modal though your application didn't go through since the app was asleep / waking up when you hit "Apply"

This probably won't happen on production since we are more dyno allocated but I think it's still worth revising.  What do you think @Pomax ? :grin: